docs/functions.md

```diff
@@ -4,54 +4,143 @@ title: Functions
 
 ```rust
-pub fn construct_js_function(mut cx: FunctionContext) -> JsResult<JsNumber> {
-    let f = cx.argument::<JsFunction>(0)?;
-    let zero = cx.number(0.0);
-    let o = f.construct(&mut cx, vec![zero])?;
-    let get_utc_full_year_method = o.get(&mut cx, "getUTCFullYear")?.downcast::<JsFunction>().or_throw(&mut cx)?;
-    let args: Vec<Handle<JsValue>> = vec![];
-    get_utc_full_year_method.call(&mut cx, o.upcast::<JsValue>(), args)?.downcast::<JsNumber>().or_throw(&mut cx)
+fn create_book(mut cx: FunctionContext) -> JsResult<JsObject> {
+    let title = cx.argument::<JsString>(0)?;
+    let author = cx.argument::<JsString>(1)?;
+    let year = cx.argument::<JsNumber>(2)?;
+
+    let obj = cx.empty_object();
+
+    obj.set(&mut cx, "title", title)?;
+    obj.set(&mut cx, "author", author)?;
+    obj.set(&mut cx, "year", year)?;
+
+    Ok(obj)
+}
+
+#[neon::main]
+pub fn main(mut cx: ModuleContext) -> NeonResult<()> {
+    cx.export_function("createBook", create_book)?;
+    Ok(())
+}
+```
+
+If any of the checks fails, the `createBook` function throws a [`TypeError`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError):
+
+```javascript
+const { createBook } = require('./index');
+
+try {
+  createBook(null, null, null);
+} catch (e) {
+  console.log(e); // TypeError
 }
 ```
 
-## Returning Functions
+## Optional Arguments
+
+The [`FunctionContext::argument_opt()`](https://docs.rs/neon/latest/neon/context/struct.CallContext.html#method.argument_opt) method makes it possible to extract an optional argument. This example creates an entry in a resume's job history, where the end year may not be present (indicating the person's current job):
 
 ```rust
-pub fn return_js_function(mut cx: FunctionContext) -> JsResult<JsFunction> {
-    JsFunction::new(&mut cx, add1)
+fn create_job(mut cx: FunctionContext) -> JsResult<JsObject> {
+    let company = cx.argument::<JsString>(0)?;
+    let title = cx.argument::<JsString>(1)?;
+    let start_year = cx.argument::<JsNumber>(2)?;
+    let end_year = cx.argument_opt(3);
+
+    let obj = cx.empty_object();
+
+    obj.set(&mut cx, "company", company)?;
+    obj.set(&mut cx, "title", title)?;
+    obj.set(&mut cx, "startYear", start_year)?;
+
+    if let Some(end_year) = end_year {
+        obj.set(&mut cx, "endYear", end_year)?;
+    } else {
+        let null = cx.null();
+        obj.set(&mut cx, "endYear", null)?;
+    }
+
+    Ok(obj)
 }
 ```
+
+## Calling Functions
+
+You can call a JavaScript function from Rust with [`JsFunction::call()`](https://docs.rs/neon/latest/neon/types/struct.JsFunction.html#method.call). This example extracts the [`parseInt`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/parseInt) function and calls it on a string:
+
+```rust
+let global = cx.global();
+let func: Handle<JsFunction> = global.get(&mut cx, "parseInt")?;
```

Comment:
This is missing a call to `downcast_or_throw`. `get` always returns a `JsValue`.

Reply:
Good catch. This makes me wish `Object::get()` returned a generic type and automatically did the downcast.